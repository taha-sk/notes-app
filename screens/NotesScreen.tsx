import React from 'react';
import { FlatList, Pressable, StyleSheet, Text, View } from 'react-native';
import { AntDesign } from '@expo/vector-icons'; 
import { NoteItem, RootStackParamList } from '../types';
import { v4 as uuidv4 } from 'uuid';
import { StackScreenProps } from '@react-navigation/stack';
import AsyncStorage from '@react-native-async-storage/async-storage';

export default function NotesScreen({navigation} : StackScreenProps<RootStackParamList, 'Notes'>) {
  
  const init: NoteItem[] =[];
  const [notes, setNotes] = React.useState(init);
  const [err, setErr] = React.useState('');

  const initializeData = async () => {
    try {
      const jsonValue = await AsyncStorage.getItem('@notes');
      console.log(jsonValue);
      if(jsonValue != null){
        const storedNotes: NoteItem[] = JSON.parse(jsonValue);
        setNotes(storedNotes);
      }
    } catch(e) {
      const errorMessage = (e as Error).message
      setErr('Error occured during data initialization: ' + errorMessage);
    }
  };

  const clearAll = async () => {
    try {
      await AsyncStorage.clear()
    } catch(e) {
      // clear error
    }
  
    console.log('Done.')
  };

  React.useEffect(() => {initializeData()}, []);

  return (
    <View style={styles.container}>
      <View style={styles.addNoteSign}>
        <Pressable onPress={() => navigation.navigate('EditNote', { noteId: ''})}>
          <AntDesign name="plus" size={24} color="black" />
        </Pressable>
      </View>
      <FlatList 
        data={notes}
        renderItem={({item}) => <Text>{item.note}</Text>}>
      </FlatList>
      <View>
          <Text>{JSON.stringify(notes.length) + ' Notes'}</Text>
          {err.length > 0 && <Text>{err}</Text>}
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    padding: 20
  },
  addNoteSign: {
    flexDirection: "row-reverse"
  }
});
