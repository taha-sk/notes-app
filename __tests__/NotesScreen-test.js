import * as React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';

import NotesScreen from '../screens/NotesScreen';

describe('<NotesScreen />', () => {

  it('renders correctly', () => {
    const navigation = { addListener: jest.fn(), navigate: jest.fn()};
    const tree = renderer.create(<NotesScreen navigation={navigation}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  it('has 3 child', () => {
    const navigation = { addListener: jest.fn(), navigate: jest.fn()};
    const tree = renderer.create(<NotesScreen navigation={navigation}/>).toJSON();
    expect(tree.children.length).toBe(3);
  });  

  it('initializes correctly', () => {
    const navigation = { addListener: jest.fn(), navigate: jest.fn()};
    const { getByText } = render(<NotesScreen navigation={navigation} />);
    const numberOfNotes = getByText('0 Notes');
    expect(numberOfNotes).toBeTruthy();
  });
  
  it('add button navigates correctly', () => {
    const navigation = { addListener: jest.fn(), navigate: jest.fn()};
    const { getByTestId } = render(<NotesScreen navigation={navigation} />);
    const addBtn = getByTestId('add-btn');
    fireEvent.press(addBtn);
    expect(navigation.navigate).toHaveBeenCalledWith('EditNote', { noteId: ''});
  });

});
 