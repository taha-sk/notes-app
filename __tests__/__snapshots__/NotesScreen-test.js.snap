// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotesScreen /> renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#f0f1f6",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row-reverse",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="add-btn"
    >
      <Text />
    </View>
  </View>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text>
      0 Notes
    </Text>
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "#f0f1f6",
          },
          Object {
            "height": 20,
          },
        ]
      }
    >
      <Text>
        Clear All
      </Text>
    </View>
  </View>
</View>
`;
